File: frontend/src-tauri/src/services/feedback.rs
//! Feedback service for submitting bug reports and feature requests via Cloudflare Worker proxy.
//!
//! The worker holds the GitHub token server-side. The desktop app sends feedback
//! to the worker with an anti-abuse key, and the worker creates the GitHub issue.

use crate::models::feedback::{
    FeedbackCreate, FeedbackResponse, FeedbackStatus, FeedbackType, PendingFeedback, SystemInfo,
};
use anyhow::{Context, Result};
use reqwest::Client;
use serde::{Deserialize, Serialize};
use std::path::PathBuf;

/// Worker base URL for the Grafyn updater/feedback proxy.
const WORKER_URL: &str = "https://grafyn-updater.grafyn-updater.workers.dev";

/// Anti-abuse key sent via X-Feedback-Key header.
/// This is NOT a secret — it only prevents casual abuse of the feedback endpoint.
/// The worst case if extracted is spam issues on a public repo.
const FEEDBACK_KEY: &str = "gfyn-fb-a7x9k2m4";

/// Service for handling feedback submission via Cloudflare Worker proxy
#[derive(Debug, Clone)]
pub struct FeedbackService {
    client: Client,
    store_path: PathBuf,
}

impl FeedbackService {
    /// Create a new feedback service. No environment variables needed.
    pub fn new(store_path: PathBuf) -> Self {
        // Create the pending feedback directory
        std::fs::create_dir_all(&store_path).ok();

        Self {
            client: Client::new(),
            store_path,
        }
    }

    /// Get the current status of the feedback service
    pub fn get_status(&self) -> FeedbackStatus {
        let pending = self.get_pending().unwrap_or_default();

        FeedbackStatus {
            configured: true,
            pending_count: pending.len(),
            message: if pending.is_empty() {
                "Feedback service ready".to_string()
            } else {
                format!("{} feedback item(s) pending submission", pending.len())
            },
        }
    }

    /// Get system information for the current platform
    pub fn get_system_info(&self, current_page: Option<String>) -> SystemInfo {
        let platform = format!(
            "{} {}",
            std::env::consts::OS,
            std::env::consts::ARCH
        );

        SystemInfo {
            platform,
            app_version: env!("CARGO_PKG_VERSION").to_string(),
            runtime: "tauri-desktop".to_string(),
            current_page,
        }
    }

    /// Submit feedback - sends to worker proxy or queues if offline
    pub async fn submit(&self, mut feedback: FeedbackCreate) -> Result<FeedbackResponse> {
        // Validate input
        feedback.validate().map_err(|e| anyhow::anyhow!(e))?;

        // Add system info if requested
        if feedback.include_system_info && feedback.system_info.is_none() {
            feedback.system_info = Some(self.get_system_info(None));
        }

        // Check connectivity
        if !self.is_online().await {
            self.queue_feedback(feedback)?;
            return Ok(FeedbackResponse::queued());
        }

        // Submit to worker
        match self.submit_to_worker(&feedback).await {
            Ok(response) => Ok(response),
            Err(e) => {
                // Queue on failure
                log::warn!("Failed to submit feedback, queueing: {}", e);
                self.queue_feedback(feedback)?;
                Ok(FeedbackResponse::queued())
            }
        }
    }

    /// Check if we can reach the worker
    async fn is_online(&self) -> bool {
        self.client
            .get(WORKER_URL)
            .send()
            .await
            .map(|r| r.status().is_success())
            .unwrap_or(false)
    }

    /// Submit feedback to the Cloudflare Worker proxy
    async fn submit_to_worker(&self, feedback: &FeedbackCreate) -> Result<FeedbackResponse> {
        // Build issue body
        let body = self.format_issue_body(feedback);

        // Determine labels based on feedback type
        let labels = match feedback.feedback_type {
            FeedbackType::Bug => vec!["bug", "user-feedback"],
            FeedbackType::Feature => vec!["enhancement", "user-feedback"],
            FeedbackType::General => vec!["feedback", "user-feedback"],
        };

        let request_body = WorkerFeedbackRequest {
            title: feedback.title.clone(),
            body,
            labels,
        };

        let response = self
            .client
            .post(format!("{}/feedback", WORKER_URL))
            .header("X-Feedback-Key", FEEDBACK_KEY)
            .json(&request_body)
            .send()
            .await
            .context("Failed to send request to feedback proxy")?;

        if !response.status().is_success() {
            let error_text = response.text().await.unwrap_or_default();
            return Err(anyhow::anyhow!("Feedback proxy error: {}", error_text));
        }

        let issue: WorkerFeedbackResponse = response
            .json()
            .await
            .context("Failed to parse feedback proxy response")?;

        Ok(FeedbackResponse::success(issue.number, issue.html_url))
    }

    /// Format the GitHub issue body with feedback details
    fn format_issue_body(&self, feedback: &FeedbackCreate) -> String {
        let type_emoji = match feedback.feedback_type {
            FeedbackType::Bug => "🐛",
            FeedbackType::Feature => "💡",
            FeedbackType::General => "💬",
        };

        let type_label = match feedback.feedback_type {
            FeedbackType::Bug => "Bug Report",
            FeedbackType::Feature => "Feature Request",
            FeedbackType::General => "General Feedback",
        };

        let mut body = format!(
            "## {} {}\n\n{}\n\n",
            type_emoji, type_label, feedback.description
        );

        if let Some(ref system_info) = feedback.system_info {
            body.push_str("---\n\n");
            body.push_str("### System Information\n\n");
            body.push_str(&format!("- **Platform:** {}\n", system_info.platform));
            body.push_str(&format!("- **App Version:** {}\n", system_info.app_version));
            body.push_str(&format!("- **Runtime:** {}\n", system_info.runtime));
            if let Some(ref page) = system_info.current_page {
                body.push_str(&format!("- **Current Page:** {}\n", page));
            }
        }

        body.push_str("\n---\n*Submitted via Grafyn Desktop App*");

        body
    }

    /// Queue feedback for later submission (offline mode)
    fn queue_feedback(&self, feedback: FeedbackCreate) -> Result<()> {
        let pending = PendingFeedback::new(feedback);
        let file_path = self.store_path.join(format!("{}.json", pending.id));

        let json = serde_json::to_string_pretty(&pending)
            .context("Failed to serialize pending feedback")?;

        std::fs::write(&file_path, json).context("Failed to write pending feedback file")?;

        log::info!("Queued feedback: {}", pending.id);
        Ok(())
    }

    /// Get all pending feedback items
    pub fn get_pending(&self) -> Result<Vec<PendingFeedback>> {
        let mut pending = Vec::new();

        if !self.store_path.exists() {
            return Ok(pending);
        }

        for entry in std::fs::read_dir(&self.store_path)? {
            let entry = entry?;
            let path = entry.path();

            if path.extension().map(|e| e == "json").unwrap_or(false) {
                if let Ok(content) = std::fs::read_to_string(&path) {
                    if let Ok(item) = serde_json::from_str::<PendingFeedback>(&content) {
                        pending.push(item);
                    }
                }
            }
        }

        // Sort by creation time, oldest first
        pending.sort_by(|a, b| a.created_at.cmp(&b.created_at));

        Ok(pending)
    }

    /// Retry submitting pending feedback items
    pub async fn retry_pending(&self) -> Result<Vec<FeedbackResponse>> {
        if !self.is_online().await {
            return Ok(vec![]);
        }

        let pending = self.get_pending()?;
        let mut results = Vec::new();

        for mut item in pending {
            match self.submit_to_worker(&item.feedback).await {
                Ok(response) => {
                    // Remove from queue on success
                    let file_path = self.store_path.join(format!("{}.json", item.id));
                    std::fs::remove_file(&file_path).ok();
                    log::info!("Successfully submitted queued feedback: {}", item.id);
                    results.push(response);
                }
                Err(e) => {
                    // Update retry count
                    item.retry_count += 1;
                    let file_path = self.store_path.join(format!("{}.json", item.id));
                    if let Ok(json) = serde_json::to_string_pretty(&item) {
                        std::fs::write(&file_path, json).ok();
                    }
                    log::warn!(
                        "Failed to retry feedback {} (attempt {}): {}",
                        item.id,
                        item.retry_count,
                        e
                    );
                    results.push(FeedbackResponse::error(format!(
                        "Failed to submit: {}",
                        e
                    )));
                }
            }
        }

        Ok(results)
    }

    /// Clear a specific pending feedback item
    pub fn clear_pending(&self, id: &str) -> Result<()> {
        // Validate ID to prevent path traversal
        if id.is_empty() || id.contains('/') || id.contains('\\') || id.contains("..") {
            anyhow::bail!("Invalid feedback ID: {}", id);
        }
        let file_path = self.store_path.join(format!("{}.json", id));
        if file_path.exists() {
            std::fs::remove_file(&file_path).context("Failed to remove pending feedback file")?;
        }
        Ok(())
    }
}

// Worker proxy request/response types

#[derive(Debug, Serialize)]
struct WorkerFeedbackRequest {
    title: String,
    body: String,
    labels: Vec<&'static str>,
}

#[derive(Debug, Deserialize)]
struct WorkerFeedbackResponse {
    number: u64,
    html_url: String,
}

File: backend/app/services/feedback.py
"""Feedback service for submitting bug reports and feature requests to GitHub"""

import httpx
import logging
from typing import Optional
from app.config import get_settings
from app.models.feedback import (
    FeedbackCreate,
    FeedbackResponse,
    FeedbackStatus,
    FeedbackType,
    SystemInfo,
)

logger = logging.getLogger(__name__)

GITHUB_API_URL = "https://api.github.com"


class FeedbackService:
    """Service for handling feedback submission to GitHub Issues"""

    def __init__(self):
        settings = get_settings()
        self.repo = settings.github_feedback_repo
        self.token = settings.github_feedback_token
        self._client: Optional[httpx.AsyncClient] = None

    async def _get_client(self) -> httpx.AsyncClient:
        """Get or create HTTP client"""
        if self._client is None or self._client.is_closed:
            self._client = httpx.AsyncClient(
                timeout=httpx.Timeout(30.0),
                headers={
                    "User-Agent": "Grafyn-Backend",
                    "Accept": "application/vnd.github+json",
                    "X-GitHub-Api-Version": "2022-11-28",
                },
            )
        return self._client

    def is_configured(self) -> bool:
        """Check if the service is properly configured"""
        return bool(self.repo and self.token)

    def get_status(self) -> FeedbackStatus:
        """Get the current status of the feedback service"""
        if not self.is_configured():
            return FeedbackStatus(
                configured=False,
                pending_count=0,
                message="Feedback service not configured. Set GITHUB_FEEDBACK_REPO and GITHUB_FEEDBACK_TOKEN.",
            )

        return FeedbackStatus(
            configured=True,
            pending_count=0,  # Web backend doesn't support offline queue
            message="Feedback service ready",
        )

    def get_system_info(self, current_page: Optional[str] = None) -> SystemInfo:
        """Get system information for the feedback form"""
        return SystemInfo(
            platform="Web Browser",
            app_version="1.0.0",
            runtime="python-fastapi",
            current_page=current_page,
        )

    async def submit(self, feedback: FeedbackCreate) -> FeedbackResponse:
        """Submit feedback to GitHub Issues"""
        if not self.is_configured():
            return FeedbackResponse.error_response(
                "Feedback service not configured. Please set GITHUB_FEEDBACK_REPO and GITHUB_FEEDBACK_TOKEN."
            )

        try:
            # Add system info if requested but not provided
            if feedback.include_system_info and feedback.system_info is None:
                feedback.system_info = self.get_system_info()

            return await self._submit_to_github(feedback)

        except httpx.ConnectError:
            logger.warning("Failed to connect to GitHub API")
            return FeedbackResponse.error_response(
                "Unable to connect to GitHub. Please check your internet connection."
            )
        except httpx.TimeoutException:
            logger.warning("GitHub API request timed out")
            return FeedbackResponse.error_response(
                "Request to GitHub timed out. Please try again."
            )
        except Exception as e:
            logger.exception("Unexpected error submitting feedback")
            return FeedbackResponse.error_response(f"An unexpected error occurred: {str(e)}")

    async def _submit_to_github(self, feedback: FeedbackCreate) -> FeedbackResponse:
        """Submit feedback directly to GitHub Issues API"""
        if "/" not in self.repo:
            return FeedbackResponse.error_response(
                "Invalid repo format, expected 'owner/repo'"
            )

        owner, repo = self.repo.split("/", 1)
        client = await self._get_client()

        # Build issue body
        body = self._format_issue_body(feedback)

        # Determine labels based on feedback type
        labels = self._get_labels(feedback.feedback_type)

        request_data = {
            "title": feedback.title,
            "body": body,
            "labels": labels,
        }

        response = await client.post(
            f"{GITHUB_API_URL}/repos/{owner}/{repo}/issues",
            json=request_data,
            headers={"Authorization": f"Bearer {self.token}"},
        )

        if response.status_code == 201:
            data = response.json()
            return FeedbackResponse.success_response(
                issue_number=data["number"],
                issue_url=data["html_url"],
            )
        else:
            error_text = response.text
            logger.error(f"GitHub API error: {response.status_code} - {error_text}")
            return FeedbackResponse.error_response(
                f"Failed to create GitHub issue: {response.status_code}"
            )

    def _format_issue_body(self, feedback: FeedbackCreate) -> str:
        """Format the GitHub issue body with feedback details"""
        type_emoji = {
            FeedbackType.BUG: "🐛",
            FeedbackType.FEATURE: "💡",
            FeedbackType.GENERAL: "💬",
        }

        type_label = {
            FeedbackType.BUG: "Bug Report",
            FeedbackType.FEATURE: "Feature Request",
            FeedbackType.GENERAL: "General Feedback",
        }

        emoji = type_emoji.get(feedback.feedback_type, "💬")
        label = type_label.get(feedback.feedback_type, "General Feedback")

        body = f"## {emoji} {label}\n\n{feedback.description}\n\n"

        if feedback.system_info:
            body += "---\n\n"
            body += "### System Information\n\n"
            body += f"- **Platform:** {feedback.system_info.platform}\n"
            body += f"- **App Version:** {feedback.system_info.app_version}\n"
            body += f"- **Runtime:** {feedback.system_info.runtime}\n"
            if feedback.system_info.current_page:
                body += f"- **Current Page:** {feedback.system_info.current_page}\n"

        body += "\n---\n*Submitted via Grafyn Web App*"

        return body

    def _get_labels(self, feedback_type: FeedbackType) -> list[str]:
        """Get GitHub labels based on feedback type"""
        label_map = {
            FeedbackType.BUG: ["bug", "user-feedback"],
            FeedbackType.FEATURE: ["enhancement", "user-feedback"],
            FeedbackType.GENERAL: ["feedback", "user-feedback"],
        }
        return label_map.get(feedback_type, ["feedback", "user-feedback"])

    async def close(self):
        """Close the HTTP client"""
        if self._client and not self._client.is_closed:
            await self._client.aclose()
            self._client = None
